### PR TITLE
DirichletBC with dolfinx.Constant

### LINKED ISSUE #507: Allow DirichletBC to be set with a Constant
At present, a DirichletBC can only be set with a `Function`, but it is often just a `Constant` value.

### PR DESCRIPTION
**Pull request by jorgensd**
Resolves #507

### CODE REVIEW

File: cpp/dolfinx/fem/DirichletBC.h [hunk 1]
```diff
@@ -192,7 +236,7 @@ class DirichletBC
   {
     assert(_dofs0.size() == _dofs1_g.size());
     assert(_function_space);
-    assert(_g);
+    assert(g);
```
Comment (jorgensd) [inline cpp/dolfinx/fem/DirichletBC.h:239]:
```suggestion
    assert(_g);
```

File: cpp/dolfinx/fem/DirichletBC.cpp [hunk 2]
```diff
@@ -139,7 +139,9 @@ get_remote_bcs1(const common::IndexMap& map,
 /// condition detected by another process
 ///
 /// @param[in] map0 The IndexMap with the dof layout
+/// @param[in] bs0 The block size of the index map
```
Comment (jhale) [inline cpp/dolfinx/fem/DirichletBC.cpp:142]: The block size of the index map map0

File: cpp/dolfinx/fem/DirichletBC.cpp [hunk 3]
```diff
@@ -139,7 +139,9 @@ get_remote_bcs1(const common::IndexMap& map,
...
 /// @param[in] map1 The IndexMap with the dof layout
+/// @param[in] bs1 The block size of the index map
```
Comment (jhale) [inline cpp/dolfinx/fem/DirichletBC.cpp:144]: The block size of the index map map1

File: cpp/dolfinx/fem/DirichletBC.cpp [hunk 4]
```diff
@@ -271,6 +273,114 @@ get_remote_bcs2(const common::IndexMap& map0, int bs0,
...
+
+  // FIXME: check that dofs is sorted?
+  // Build vector of local dof indicies that have been marked by
+  // another process
+  const std::array<std::int64_t, 2> range = map.local_range();
+  const std::vector<std::int64_t>& ghosts = map.ghosts();
+
+  // Build map from ghost to local position
+  std::vector<std::pair<std::int64_t, std::int32_t>> global_local_ghosts;
+  const std::int32_t local_size = range[1] - range[0];
```
Comment (jhale) [inline cpp/dolfinx/fem/DirichletBC.cpp:357]: `range` contains `int64_t` and is cast to `int32_t`?

File: python/test/unit/fem/test_bcs.py [hunk 5]
```diff
@@ -97,5 +100,107 @@ def test_overlapping_bcs():
...
+
+    u_f = Function(V)
+    set_bc(u_f.vector, [bc_f0, bc_f1])
+
+    u_c = Function(V)
+    set_bc(u_c.vector, [bc_c])
+    assert(np.allclose(u_f.vector.array, u_c.vector.array))
+
+
+def test_sub_constant_bc():
```
Comment (jhale) [inline python/test/unit/fem/test_bcs.py:173]: Parameterise over 1D/2D/3D meshes?

File: python/test/unit/fem/test_bcs.py [hunk 6]
```diff
@@ -97,5 +100,107 @@ def test_overlapping_bcs():
...
+
+    u_f = Function(V)
+    set_bc(u_f.vector, [bc_f])
+
+    u_c = Function(V)
+    set_bc(u_c.vector, [bc_c])
+    assert(np.allclose(u_f.vector.array, u_c.vector.array))
+
+
+def test_vector_constant_bc():
```
Comment (jhale) [inline python/test/unit/fem/test_bcs.py:136]: Parameterise over 1D/2D/3D meshes?

File: python/test/unit/fem/test_bcs.py [hunk 7]
```diff
@@ -97,5 +100,107 @@ def test_overlapping_bcs():
 
     if len(dof_corner) > 0:
         with b.localForm() as b_loc:
-            print(b_loc[dof_corner[0]])
             assert b_loc[dof_corner[0]] == 123.456
+
+
+def test_constant_bc():
```
Comment (jhale) [inline python/test/unit/fem/test_bcs.py:106]: Parameterise over 1D/2D/3D meshes?

File: cpp/dolfinx/fem/DirichletBC.cpp [hunk 8]
```diff
@@ -427,56 +537,114 @@ fem::locate_dofs_topological(const fem::FunctionSpace& V, int dim,
...
+      // were found by other processes, e.g. a vertex dof on this process
+      // that has no connected facets on the boundary.
+      const std::vector dofs_remote = get_remote_bcs1_sub(
+          *V.dofmap()->index_map, V.dofmap()->index_map_bs(), dofs);
+
+      // Add received bc indices to dofs_local
+      dofs.insert(dofs.end(), dofs_remote.begin(), dofs_remote.end());
+
+      // Remove duplicates and sort
+      std::sort(dofs.begin(), dofs.end());
```
Comment (jhale) [inline cpp/dolfinx/fem/DirichletBC.cpp:644]: It it necessary to do the sort and erase twice?

File: cpp/dolfinx/fem/DirichletBC.cpp [hunk 9]
```diff
@@ -271,6 +273,114 @@ get_remote_bcs2(const common::IndexMap& map0, int bs0,
...
+  }
+
+  // Compute displacements for data to receive. Last entry has total
+  // number of received items.
+  MPI_Wait(&request, MPI_STATUS_IGNORE);
+  std::vector<int> disp(num_neighbors + 1, 0);
+  std::partial_sum(num_dofs_recv.begin(), num_dofs_recv.end(),
+                   std::next(disp.begin()));
+
+  // NOTE: we could use MPI_Neighbor_alltoallv to send only to relevant
```
Comment (jhale) [inline cpp/dolfinx/fem/DirichletBC.cpp:338]: Which line is this comment referring to?

Comment (jorgensd) [reply cpp/dolfinx/fem/DirichletBC.cpp:338]: Good question, i think this is a remainder from my copy-paste of the similar functions `get_remote_bcs` and `get_remote_bcs2` that has the same comment.

File: cpp/dolfinx/fem/DirichletBC.cpp (hunk 8 above)
Comment (jorgensd) [reply cpp/dolfinx/fem/DirichletBC.cpp:644]: Good question, this is done in the other implementations of `locate_dofs_topological` as well.

File: cpp/dolfinx/fem/DirichletBC.cpp (hunk 4 above)
Comment (jorgensd) [reply cpp/dolfinx/fem/DirichletBC.cpp:357]: Follows the same structure as the other codes, not that this is a good justification.

I will look into how to rewrite all the get_remote_bcs functions.

Comment (jorgensd) [reply cpp/dolfinx/fem/DirichletBC.cpp:397]: @jhale  This is exactly the same thing we do internally in IndexMap: https://github.com/FEniCS/dolfinx/blob/main/cpp/dolfinx/common/IndexMap.cpp#L438

**Comment by jorgensd** [comment]
Superseeded by #1883

**Pull request closed without merging**